core/arch/instrlist.c

```diff
@@ -481,15 +481,19 @@ instrlist_encode_to_copy(dcontext_t *dcontext, instrlist_t *ilist, byte *copy_pc
     instr_t *inst;
     int len = 0;
 #ifdef ARM
-    /* XXX i#1734: reset encdoe state to avoid any stale encode state
+    /* XXX i#1734: reset encode state to avoid any stale encode state
      * or dangling pointer.
      */
     if (instr_get_isa_mode(instrlist_first(ilist)) == DR_ISA_ARM_THUMB)
         encode_reset_it_block(dcontext);
 #endif
-    if (has_instr_jmp_targets || max_pc != NULL) {
+    if (IF_DEBUG_ELSE(true, has_instr_jmp_targets) || max_pc != NULL) {
         /* must set note fields first with offset, or compute length */
         for (inst = instrlist_first(ilist); inst; inst = instr_get_next(inst)) {
+            if (IF_DEBUG_ELSE(!has_instr_jmp_targets && instr_is_label(inst), false)) {
+                SYSLOG_INTERNAL_WARNING(
+                        "has_instr_jmp_targets was unset but a label was found");
+            }
```

Comment:
Done w/ adding the `CLIENT_ASSERT` in x86/encode.c, but for some reason I am unable to perform the `DOCHECK` in `instrlist_encode_to_copy`, here.

If I try to call `instr_num_srcs` (i.e. to check for an inst opnd as in your comment above) I get a segfault because `dcontext` is NULL. The stacktrace also is puzzling, because it seems that `instr_num_srcs` is trying to decode the instruction and perform some heap allocation:

```
#0  0x0000555555705191 in heap_alloc (dcontext=0x0, size=0x20, which=ACCT_IR) at /home/piazzt/Build/dynamorio/core/heap.c:3589
#1  0x00005555557aa610 in instr_set_num_opnds (dcontext=0x0, instr=0x4e4689f8, instr_num_dsts=0x2, instr_num_srcs=0x1)
    at /home/piazzt/Build/dynamorio/core/arch/instr_shared.c:535
#2  0x00005555557be553 in decode_common (dcontext=0x0, pc=0x4e468c68 "\234\274\274\274\274\274\274\274",
    orig_pc=0x4e468c68 "\234\274\274\274\274\274\274\274", instr=0x4e4689f8)
    at /home/piazzt/Build/dynamorio/core/arch/x86/decode.c:2070
#3  0x00005555557bea9e in decode (dcontext=0x0, pc=0x4e468c68 "\234\274\274\274\274\274\274\274", instr=0x4e4689f8)
    at /home/piazzt/Build/dynamorio/core/arch/x86/decode.c:2145
#4  0x00005555557ac5ca in instr_decode (dcontext=0x0, instr=0x4e4689f8)
    at /home/piazzt/Build/dynamorio/core/arch/instr_shared.c:1529
#5  0x00005555557ac6a0 in instr_decode_with_current_dcontext (instr=0x4e4689f8)
    at /home/piazzt/Build/dynamorio/core/arch/instr_shared.c:1554
#6  0x00005555557b8a04 in instrlist_encode_to_copy (dcontext=0xffffffffffffffff, ilist=0x7fffffffd000,
    copy_pc=0x4e47eeaa "\234", '\253' <repeats 199 times>..., final_pc=0x4e47eeaa "\234", '\253' <repeats 199 times>...,
    max_pc=0x0, has_instr_jmp_targets=0x0) at /home/piazzt/Build/dynamorio/core/arch/instrlist.c:514
#7  0x00005555557b8a87 in instrlist_encode (dcontext=0xffffffffffffffff, ilist=0x7fffffffd000, pc=0x4e47edc0 "eH\211<%\030",
    has_instr_jmp_targets=0x0) at /home/piazzt/Build/dynamorio/core/arch/instrlist.c:533
#8  0x0000555555791862 in emit_fcache_return_shared (dcontext=0xffffffffffffffff, code=0x4e47d000, pc=0x4e47edc0 "eH\211<%\030")
    at /home/piazzt/Build/dynamorio/core/arch/emit_utils_shared.c:2522
#9  0x0000555555784848 in shared_gencode_emit (gencode=0x4e47d000, x86_mode=0x0)
    at /home/piazzt/Build/dynamorio/core/arch/arch.c:328
#10 0x0000555555784f5a in shared_gencode_init (gencode_mode=GENCODE_X64) at /home/piazzt/Build/dynamorio/core/arch/arch.c:545
#11 0x0000555555785213 in arch_init () at /home/piazzt/Build/dynamorio/core/arch/arch.c:723
#12 0x00005555555e4c02 in dynamorio_app_init () at /home/piazzt/Build/dynamorio/core/dynamo.c:543
#13 0x000055555585d0bc in privload_early_inject (sp=0x7fffffffe360, old_libdr_base=0x555555580a58 "\250\204`q",
    old_libdr_size=0x7fffffffe1a0) at /home/piazzt/Build/dynamorio/core/unix/loader.c:1838
#14 0x0000555555811b38 in reloaded_xfer () at /home/piazzt/Build/dynamorio/core/arch/x86/x86.asm:1186
#15 0x0000000000000001 in ?? ()
#16 0x00007fffffffe5ed in ?? ()
#17 0x0000000000000000 in ?? ()
```

Do you have any suggestions? Should I just go back to checking for labels instead of for opnd inst's?